---
title: "R Notebook"
#output: html_notebook
---

# Loading data
## Load data into R
```{r}
# set root directory
root = "/home/student12/project/"

# Load the metadata, use paste0() to concatenate two strings
sample_metadata <- read_tsv(paste0(root,"data/sample_metadata.txt"))

```

# QC Preprocessing
Trim to minimal 35 bases and 35 quality score: **Trimmomatic v0.22**

## Packages and software
```{bash}
# Initiate conda
source ~/.bashrc

# Using the sequencing conda environment containing:
# conda install samtools bowtie2 breseq abyss bcftools wgsim emboss tree

conda install -c bioconda trimmomatic

# Not mandatory: install fastqc to view fastq files in browser
conda install -c bioconda fastqc

```



## Trimming and quality control
```{bash}



```


# de novo assembly
**ABySS v1.3.7**


# Disassembly BAM files
**SamToFastq.jar** and **SAMtools**


# Alignment
(reference, de novo concensus) with very sensitive local alignment:
**Bowtie2 v2.2.4**



# Compress SAM file and sort by coordinate
**Samtool view** and **SortSam** using:
**SAMtools v1.1** and **Picard-tools v1.128**


# Remove PCR duplicates
**MarkDuplicates** and **Picard-tools v1.128**


# Variant detection
**V-phaser2 v2.0**


# Shannon entropy
**RF** and **bam2R** using:
**deepSNV v1.16.0**

```{r}
new_entropy = function(x){
  x[1:5]
  new_entropy = -sum(x*log(x), na.rm = TRUE)
  
  return(new_entropy)}
```


