---
title: "Result visualization"
output: html_document
---


## Instructions :

This notebook presents all the codes used to analyze and plot the data obtained previously. The files used have been previously moved into the folder results.

This script is divided into two sections, and each section correspond to a figure of the repport :

* **Shannon Entropy** :
  + Mean Shannon Entropy per passage per strain (figure 2)
  + Mean Shannon Entropy and Mean Coverage (figure 3)
* **Variants detection** :
  + SNVs frequencies per passage per strain (figure 4)
  + SNVs mapping across the genome (figure 6)
  + Mean Coverage and Total ammount of SNVs (figure 5)
  
  
```{r, include = FALSE}

# All the libraries required :
library(tidyverse)
library(ggrepel)
library(ggprism)

```


## Shannon Entropy 
#### Mean Shannon Entropy per passage per strain

Description of the variables used :

* **shannon_entropies** : a data frame gathering for each passage of each strain (column) the calculated shannon entropy for each position in the genome (rows)
* **mean_entropies** : list of the mean entropy for each passage/strain 
* **se** : list of the standard errors 
* **plot_WT** : a data frame that associates the mean entropy and the standard error for each passage of the WT strain
* **plot_10D** : a data frame that associates the mean entropy and the standard error for each passage of the 10D strain

```{r}

# List the data
path_to_shannon_files = "~student12/project/results/shannon_entropy/"
file_list = list.files(path_to_shannon_files, pattern = "shannon_" ,full.names = TRUE)

# Load the data in the data frame
column_names = c()
shannon_entropies = data.frame()

for (file in file_list){
  column_names = cbind(column_names, tools::file_path_sans_ext(basename(file)))
  shannon_entropies = rbind(shannon_entropies, t(read.table(file)))
}

# Needing to transpose data and rename columns
shannon_entropies = as.data.frame(t(as.matrix(shannon_entropies)))
colnames(shannon_entropies) = column_names
```

```{r}

# Helpful function to calculate standard error of the mean
standard_error = function(x) sd(x) / sqrt(length(x))

# Calculate mean entropy and standard error of the mean
mean_entropies = sapply(shannon_entropies, mean)
se = sapply(shannon_entropies, standard_error)
x = 0:5
```

```{r}

par(mfrow = c(1,2), fig=c(0.1,0.7,0.3,0.9))

#For 3'UTRΔ10_ZIKV :
plot1<-as.data.frame(cbind(x,mean_entropies[1:6]))
ggplot(data=plot1,aes(x=x,y=V2))+
  geom_point(size=3)+
  geom_line()+
  geom_errorbar(aes(ymin=V2-se[1:6],ymax=V2+se[1:6]),width=.1)+
  theme_prism()+
  labs(x="Passages",y="Mean Shanon Entropy +/- SD",title ="Mean entropy in 3'UTRΔ10 ZIKV passage 0-5")

#For WT_ZIKV :
plot2<-as.data.frame(cbind(x,mean_entropies[7:12]))
ggplot(data=plot2,aes(x=x,y=V2))+
  geom_point(size=3)+
  geom_line()+
  geom_errorbar(aes(ymin=V2-se[7:12],ymax=V2+se[7:12]),width=.1)+
  theme_prism()+
  labs(x="Passages",y="Mean Shanon Entropy +/- SD",title ="Mean entropy in WT ZIKV passage 0-5")
```


## Shannon Entropy
##### Mean Shannon Entropy and  Mean Coverage

Description of the variables used :

* **mean_coverage** : data frame containing for each passage/strain (columns) the mean coverage observed (rows)
* **cov_shannon** : data frame that associates the mean coverage with the mean entropy for each passage/strain
* **cov_ready_to_plot** : data frame with an added column which precise the strain for each couple {mean_coverage;mean_entropy}

```{r}

# List the data
path_to_SNP_files = "~student12/project/results/SNP/"
cov_file_list = list.files(path_to_SNP_files, pattern = "covplot.R" ,full.names = TRUE)

# Load the data in the folder
column_names = c()
mean_coverage = data.frame()

for (file in cov_file_list){
  column_names = cbind(column_names, tools::file_path_sans_ext(basename(file)))
  # Read the file
  source(file)
  # variable y is the vector containing the coverage in the file
  mean_coverage = rbind(mean_coverage, mean(y))
}

# Needing to transpose data and rename columns
mean_coverage = as.data.frame(t(as.matrix(mean_coverage)))
colnames(mean_coverage) = column_names

# Combine the mean coverage and mean entropies for convenience
cov_shannon = as.data.frame(cbind(t(mean_coverage), mean_entropies))

# Replace the rownames with only the strain and passage
row_names = str_replace(rownames(cov_shannon), "SNP_","") %>% str_replace(".covplot","")
rownames(cov_shannon) = row_names
colnames(cov_shannon) = c("mean_coverage","mean_entropies")  

```
```{r}

#create a new variable that contain either WT or 10
strain <-substr(row_names,1,2)

#we add those values to the data frame, to identify the samples
cov_ready_to_plot <- cbind(cov_shannon,strain)

#we can now plot the figure
ggplot(cov_ready_to_plot, aes(x=mean_coverage,y=mean_entropies))+
  geom_point(aes(shape=strain),size=3)+
  geom_smooth(method = "lm", se = FALSE)+
  theme(legend.position = "right")+
  theme_prism()+
  labs(x="Coverage Depth",y="Shannon Entropy")


```

```{r}
cor.test(cov_ready_to_plot$mean_entropies,cov_ready_to_plot$mean_coverage, method="spearman")
model_Shannon <- lm(mean_entropies~mean_coverage, data=cov_ready_to_plot)
print(model_Shannon)
```

## Variants Detection 
##### SNVs Frequencies per passage per strain

Description of the variables used :

* **all_data_frame_WT** : List of 6 data frames. Each data frame contain the position and frequency of all SNVs detected at a given passage for the WT_ZIKV.
* **all_data_frame_10D** : List of 6 data frames. Each data frame contain the position and frequency of all SNVs detected at a given passage for the 10D_ZIKV.
* **single_data_frame** : a data frame containing each position and frequency of all SNVs according to their passage and strain
```{bash, echo = FALSE, include = FALSE}

# Processing WT files to remove the two first lines and extract only the information required (position in the genome and frequency of the mutation in the viral population)
  grep -v "#" SNP_WT_P0.fdr.var.txt | cut -f 1,6  > modified_SNP_WT_P0.fdr.var.txt
  grep -v "#" SNP_WT_P1.fdr.var.txt | cut -f 1,6  > modified_SNP_WT_P1.fdr.var.txt
  grep -v "#" SNP_WT_P2.fdr.var.txt |cut -f 1,6> modified_SNP_WT_P2.fdr.var.txt
  grep -v "#" SNP_WT_P3.fdr.var.txt |cut -f 1,6> modified_SNP_WT_P3.fdr.var.txt
  grep -v "#" SNP_WT_P4.fdr.var.txt |cut -f 1,6> modified_SNP_WT_P4.fdr.var.txt
  grep -v "#" SNP_WT_P5.fdr.var.txt |cut -f 1,6> modified_SNP_WT_P5.fdr.var.txt
  
  
# Processing 10_D files to remove the two first lines and extract only the information required (position in the genome and frequency of the mutation in the viral population)
  grep -v "#" SNP_10D_P0.fdr.var.txt | cut -f 1,6  > modified_SNP_10D_P0.fdr.var.txt
  grep -v "#" SNP_10D_P1.fdr.var.txt | cut -f 1,6  > modified_SNP_10D_P1.fdr.var.txt
  grep -v "#" SNP_10D_P2.fdr.var.txt |cut -f 1,6 > modified_SNP_10D_P2.fdr.var.txt
  grep -v "#" SNP_10D_P3.fdr.var.txt |cut -f 1,6 > modified_SNP_10D_P3.fdr.var.txt
  grep -v "#" SNP_10D_P4.fdr.var.txt |cut -f 1,6 > modified_SNP_10D_P4.fdr.var.txt
  grep -v "#" SNP_10D_P5.fdr.var.txt |cut -f 1,6 > modified_SNP_10D_P5.fdr.var.txt
```

```{r}

#Creation of a data frame containing all the SNVs for each passage for each strain by listing the files of interest using list.files.

path_to_SNP_files ="/home/student12/project/results/SNP"
all_files_WT = list.files(path_to_SNP_files, pattern ="modified_SNP_WT", full.names = TRUE)
all_files_10D = list.files(path_to_SNP_files, pattern ="modified_SNP_10D", full.names = TRUE)

#Importation of every files per strain under the data frame format :

all_data_frames_WT <- lapply(all_files_WT, function(x){
  read.csv(x,header = FALSE, sep="\t") } )
all_data_frames_10D <- lapply(all_files_10D, function(x){
  read.csv(x,header = FALSE, sep="\t") } )

#Modification of the name of the columns to identify positions and frequencies.
#Creation of a new column containing the number of the passage for each SNV.

for (i in 1:6) {
  colnames(all_data_frames_WT[[i]]) <- c("Position","Frequency")
  all_data_frames_WT[[i]] <- mutate(all_data_frames_WT[[i]], Passage = i-1)
  all_data_frames_WT[[i]] <- mutate(all_data_frames_WT[[i]],Strain = "WT")
}
for (i in 1:6) {
  colnames(all_data_frames_10D[[i]]) <- c("Position","Frequency")
  all_data_frames_10D[[i]] <- mutate(all_data_frames_10D[[i]], Passage = i-1)
  all_data_frames_10D[[i]] <- mutate(all_data_frames_10D[[i]],Strain = "10D")
}

#The data frames are bind together to simplify data manipulation :
single_data_frame <- Reduce(rbind, c(all_data_frames_WT,all_data_frames_10D))
single_data_frame$Passage <- as.factor(single_data_frame$Passage)
```
```{r}

par(mfrow = c(1,2), fig=c(0.1,0.7,0.3,0.9)) 

#For the WT_ZIKV :
single_data_frame %>%
  filter(Strain =="WT")%>%
  ggplot(aes(x = Passage, y= Frequency)) + 
  geom_boxplot() + 
  geom_point(position = "jitter") +
  theme_prism()+
  labs(x = "Passage", y = "SNV Frequency (%)", title = "WT")

#For the 10D_ZIKV :
single_data_frame %>%
  filter(Strain =="10D")%>%
  ggplot(aes(x = Passage, y= Frequency)) + 
  geom_boxplot() + 
  geom_point(position = "jitter") +
  theme_prism() +
  labs(x = "Passage", y = "SNV Frequency (%)", title ="3'UTRΔ10")
```


##### Mapping Mutations in the Genome

```{r}
#The single_data_frame containing all the information required is filtered, and arranged according the position of the SNV in the genome. The repetitions of SNVs are removed with the command distinct.

mutation_of_interest <- filter(single_data_frame, Frequency >= 5) %>%
  arrange(desc(Position)) %>%
  distinct(Position, .keep_all = TRUE) %>%
  arrange(Strain)

ggplot(mutation_of_interest, aes(x=Position, y = Strain, color=Strain,label=Passage))+
  geom_point(size=3)+
  theme_prism()+
  annotate("rect", xmin = 108, xmax = 473 ,ymin =0,ymax= 3, alpha = 0.5, fill = "white")+
  annotate("rect", xmin = 474, xmax = 977 ,ymin =0,ymax= 3, alpha = 0.5, fill = "grey")+
  annotate("rect", xmin = 978, xmax = 2489 ,ymin =0,ymax= 3, alpha = 0.5, fill = "white")+
  annotate("rect", xmin = 2490, xmax = 3545 ,ymin =0,ymax= 3, alpha = 0.5, fill = "grey")+
  annotate("rect", xmin = 3546, xmax = 4223 ,ymin =0,ymax= 3, alpha = 0.5, fill = "white")+
  annotate("rect", xmin = 4224, xmax = 4613 ,ymin =0,ymax= 3, alpha = 0.5, fill = "grey")+
  annotate("rect", xmin = 4614, xmax = 6464 ,ymin =0,ymax= 3, alpha = 0.5, fill = "white")+
  annotate("rect", xmin = 6465, xmax = 6914 ,ymin =0,ymax= 3, alpha = 0.5, fill = "grey")+
  annotate("rect", xmin = 6915, xmax = 7667 ,ymin =0,ymax= 3, alpha = 0.5, fill = "white")+
  annotate("rect", xmin = 7668, xmax = 10379 ,ymin =0,ymax= 3, alpha = 0.5, fill = "grey")+
  annotate("text", x = 150, y = 2.7, label = "C")+
  annotate("text", x = 725, y = 2.5, label = "prM/M")+
  annotate("text", x = 1500, y = 2.7, label = "E")+
  annotate("text", x = 3017, y = 2.5, label = "NS1")+
  annotate("text", x = 3884, y = 2.7, label = "NS2A")+
  annotate("text", x = 4418, y = 2.5, label = "NS2B")+
  annotate("text", x = 5539, y = 2.7, label = "NS3")+
  annotate("text", x = 6689, y = 2.5, label = "NS4A")+
  annotate("text", x = 7291, y = 2.7, label = "NS4B")+
  annotate("text", x = 9023, y = 2.5, label = "NS5")+
  geom_point(size=3)+
  geom_text_repel()
```

##### Mean Coverage and Total ammount of SNV

```{r}

#The total ammount of SNV per passage per strain is calculated from the single_data_frame. The mean_coverage and the total_ammount_SNV are associated in a data frame (SNV_ammount_and_coverage) and plotted.


total_ammount_SNV = count(single_data_frame,Passage,Strain)
total_ammount_SNV = arrange(total_ammount_SNV,Strain)
SNV_ammount_and_coverage <- cbind(total_ammount_SNV,t(mean_coverage))
colnames(SNV_ammount_and_coverage) = c("Passage","Strain","n","mean_coverage")

ggplot(SNV_ammount_and_coverage, aes(x=mean_coverage,y=n))+
  geom_point(aes(shape=strain), size =3)+
  geom_smooth(method = "lm", se = FALSE)+
  theme(legend.position = "right")+
  theme_prism()+
  labs(x="Coverage Depth",y="Total of SNVs")
```

```{r}
cor.test(SNV_ammount_and_coverage$n,SNV_ammount_and_coverage$mean_coverage,method="spearman")
model_SNV <- lm(n ~ mean_coverage, data=SNV_ammount_and_coverage)
print(model_SNV)
```

